src/consensus/validation.h

```diff
@@ -106,40 +107,13 @@ class ValidationState {
     bool IsValid() const { return m_mode == MODE_VALID; }
     bool IsInvalid() const { return m_mode == MODE_INVALID; }
     bool IsError() const { return m_mode == MODE_ERROR; }
+    Result GetResult() const { return m_result; }
     std::string GetRejectReason() const { return m_reject_reason; }
     std::string GetDebugMessage() const { return m_debug_message; }
 };
 
-inline ValidationState::~ValidationState() {};
-
-class TxValidationState : public ValidationState {
-private:
-    TxValidationResult m_result;
-public:
-    bool Invalid(TxValidationResult result,
-                 const std::string &reject_reason="",
-                 const std::string &debug_message="")
-    {
-        m_result = result;
-        ValidationState::Invalid(reject_reason, debug_message);
-        return false;
-    }
-    TxValidationResult GetResult() const { return m_result; }
-};
-
-class BlockValidationState : public ValidationState {
-private:
-    BlockValidationResult m_result;
-public:
-    bool Invalid(BlockValidationResult result,
-                 const std::string &reject_reason="",
-                 const std::string &debug_message="") {
-        m_result = result;
-        ValidationState::Invalid(reject_reason, debug_message);
-        return false;
-    }
-    BlockValidationResult GetResult() const { return m_result; }
-};
+using TxValidationState = ValidationState<TxValidationResult>;
```

Comment:
I think you can do `class TxValidationState : ValidationState<TxValidationResult> { };` and still be able to use forward declarations `class TxValidationState;` fwiw.

Reply:
Good idea! Done in 0468628.